### PR TITLE
v2 Refactor

### LINKED ISSUE #55: re-work README
see comments in #53 and #54

### LINKED ISSUE #57: support skin color modifiers
Hello!

Why don't work this emoji?

https://runkit.com/embed/3gxoc1t1b4pn

### PR DESCRIPTION
**Pull request by Richienb**
Signed-off-by: Richie Bendall <richiebendall@gmail.com>

Closes #55, Closes #57, Closes #62, Closes #66, Closes #77

### CODE REVIEW

**Comment by Richienb** [comment]
@omnidan Looks like you need to remove Circle CI from trying to build this repository.

**Comment by omnidan** [comment]
@Richienb I removed Circle CI. However, Travis is failing as well due to code style problems. Please adjust the code style according to the current standard or switch to http://standardjs.com, similarly to what [omnidan/redux-undo](https://github.com/omnidan/redux-undo) uses.

**Comment by omnidan** [comment]
@Richienb thanks! CI is passing now!

can you please check if we really need `spdx-correct` and `validate-npm-package-license` as dependencies (they might be subdependencies of other packages)?

These two packages seem to be GPL-3.0 licensed (according to fossa), which might "infect" our package with GPL-3.0 as well and discourage people from using it.

**Comment by Richienb** [comment]
@omnidan They're using by `meow`, however, I don't think Sindre knows that yet.

**Comment by omnidan** [comment]
@Richienb alright, it might be fossa not detecting the license properly then. If they are Apache licensed it should be fine.

**Comment by omnidan** [comment]
I have resolved the issues manually on fossa, I think it will update on the next commit

**Comment by Richienb** [comment]
In v2, we've switched to using the official Unicode emoji names from https://unicode.org/Public/emoji/12.0/emoji-test.txt however, we **could** follow another list if need be.

**Comment by ShadowRanger** [comment]
Just putting my 2 cents in regarding this refactor.

I feel as though the use of lodash is unnecessary as it's use in this refactor can be replaced by simple native methods. Though it may not make much of a difference when it comes down to performance, I personally don't see the need to use an external library for things that can already be done easily using native JavaScript.

Additionally, is the CLI implementation in v2 really needed? I feel as though it's inclusion somewhat takes away from it being a simple library that provides emoji support for Node.js projects. Though usable, I feel most people who would download the package would not end up using it for its CLI component, and even if they did want what that provides it could be easily implemented anyways.

Other than that, good work on the refactor!

**Comment by ShadowRanger** [comment]
It doesn't look like **emoji.json** is being pushed to this v2 branch. Might want to look into that @Richienb.

**Comment by Richienb** [comment]
@ShadowRanger emoji.json as in https://www.npmjs.com/package/emoji.json

**Comment by ShadowRanger** [comment]
> @ShadowRanger emoji.json as in https://www.npmjs.com/package/emoji.json

The current version of this library comes with an **emoji.json** (https://github.com/omnidan/node-emoji/blob/master/lib/emoji.json) file, containing all of the data for the available emojis that the library can access. Your refactor seems to be missing this, and is requiring a non-existent file (https://github.com/omnidan/node-emoji/blob/v2/index.js#L5). How you want to implement this file is entirely up to you. You could add it to the library itself like is done in the current version, or you could use a dependency specifically for it. Just be sure that the library accesses the JSON data correctly.

Let me know if you would like me to elaborate on anything I have said, or if you have any trouble with it. Happy to help.

**Comment by Richienb** [comment]
@ShadowRanger I'm actually requiring the **npm module** called `emoji.json`.

**Comment by ShadowRanger** [comment]
@Richienb My bad! I completely missed that. I have no idea how I did not see that in the package.json. I could have sworn that I checked first last night, clearly I did not have a clear state of mind at the time. My apologies!

**Comment by ShadowRanger** [comment]
@Richienb Just realised that I was looking at the v2 branch of this repository, rather than all of your latest changes in this pull request from your cloned repository. Either way, still my bad for not checking the right thing! Just wanted to clarify myself.

File: index.js [hunk 1]
```diff
@@ -1 +1,138 @@
...
+}
+
+function replace (string, replacement, { removeSpaces = false } = {}) {
+  ow(string, ow.string)
+  ow(replacement, ow.any(ow.string, ow.function))
+  ow(removeSpaces, ow.boolean)
+
+  const replaceFn = typeof replacement === 'function' ? replacement : () => replacement
+
+  const chars = string.match(charRegex)
```
Comment (jamesonjlee) [inline index.js]:
```suggestion
  if (chars === null) {
    return string;
  }
```

string.match(...) can return null if there are no matches.

**Comment by JeDaYoshi** [comment]
Mind if I ask, what's the state of this PR?

**Comment by Richienb** [comment]
> Mind if I ask, what's the state of this PR?

We're almost there. Just gotta do some cleanup.

**Comment by Richienb** [comment]
What is the tradeoff we are willing to make in terms of dependencies vs extra code?

**Comment by omnidan** [comment]
@Richienb it depends. If we only use one or two functions and the library is huge, I would rather add the extra code. I think in general node-emoji is so simple that we should try not to use dependencies if possible.

**Comment by Richienb** [comment]
> @Richienb it depends. If we only use one or two functions and the library is huge, I would rather add the extra code. I think in general node-emoji is so simple that we should try not to use dependencies if possible.

In this case, we're using lots of really small dependencies.

File: package.json [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
+  "repository": "https://github.com/omnidan/node-emoji.git",
   "author": "Daniel Bugl <daniel.bugl@touchlay.com>",
-  "repository": {
-    "type": "git",
-    "url": "https://github.com/omnidan/node-emoji.git"
+  "license": "MIT",
+  "main": "index.js",
+  "files": [
+    "index.js",
+    "index.d.ts"
```
Comment (JoshuaKGoldberg) [inline package.json:11]: IME the three best ways to have TypeScript support are:

* Write the original package in TypeScript
* Include `.d.ts` files and tests in the repository
* Defer to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) for typings

Including raw `.d.ts` files in this source repo but not testing them means contributors+maintainers will all have to manually check them. That's kind of hard if you're not TypeScript-familiar yourself -- and if you are, then why not write in TypeScript? 😜 

Proposal: remove `index.d.ts` and file a followup to discuss either rewriting in TypeScript or moving to DefinitelyTyped?

File: .gitattributes [hunk 3]
```diff
@@ -0,0 +1,288 @@
+# Common settings that generally should always be used with your language specific settings
```
Comment (JoshuaKGoldberg) [inline .gitattributes:1]: I haven't seen many repos include _all_ these settings -`.gitattributes`, `.gitignore`, .. -- are they necessary in this PR?

File: .editorconfig [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+indent_style = space
+indent_size = 2
+
+[*.yml]
+indent_style = space
+indent_size = 2
```
Comment (JoshuaKGoldberg) [inline .editorconfig:13]: Does changing the formatting engine need to be in this PR? Seems like it's adding a lot.

Proposal: split it out so I can evangelize Prettier? 😄

File: package.json (hunk 2 above)
Comment (omnidan) [reply package.json:11]: I'm for rewriting it in TypeScript if you are up for it - that would definitely be the best solution 😁

File: .travis.yml [hunk 5]
```diff
@@ -1,3 +1,5 @@
 language: node_js
+
 node_js:
-  - "0.10"
+  - "lts/*"
+  - "node"
```
Comment (JoshuaKGoldberg) [inline .travis.yml:5]: Doesn't have to be in this PR, but just dropping a note - I'm down to switch over to GitHub Actions. Haven't seen a `.travis.yml` update that wasn't a removal in quite a while 😄

File: .editorconfig (hunk 4 above)
Comment (omnidan) [reply .editorconfig:13]: Personally a fan of prettier as well, so I'm fine with doing it in a separate PR with prettier! 💅

File: .gitattributes (hunk 3 above)
Comment (omnidan) [reply .gitattributes:1]: I agree, we should only have necessary files, maybe @Richienb can elaborate a bit?

File: package.json [hunk 6]
```diff
@@ -1,11 +1,23 @@
...
+  ],
+  "engines": {
+    "node": ">=10"
+  },
+  "scripts": {
+    "lint": "standard",
+    "test": "standard && ava",
```
Comment (JoshuaKGoldberg) [inline package.json:18]:
```suggestion
    "test": "ava",
```

Also, out of curiosity: any reason for `ava`? Nothing against it; I just haven't seen many projects using it.

Comment (omnidan) [reply package.json:18]: I think back then I had some issues with `jest`, which was a loooong time ago though. I think now `jest` would be the more standard and stable solution for this 🃏

File: README.md [hunk 7]
```diff
@@ -1,83 +1,250 @@
...
+
+console.log(emoji.get('unicorn'))
+//=> '🦄'
+```
+
+### emoji.which(emoji)
+
+Get an emoji name from an emoji.
+
+#### emoji
```
Comment (JoshuaKGoldberg) [inline README.md:57]: It's an accessibility oddity to have headings with the same name (and also linking then only works for one). Should look at getting each of these unique text values...

File: index.js [hunk 8]
```diff
@@ -1 +1,133 @@
...
+'use strict'
+
+const charRegex = require('char-regex')()
+const emojiRegex = require('emoji-regex')()
+const randomItem = require('random-item')
+const skinTone = require('skin-tone')
+const is = require('@sindresorhus/is')
+const { assert } = is
+
+const emojiData = Object.entries(require('emojilib').lib).map(([name, { char: emoji }]) => [name, emoji])
```
Comment (JoshuaKGoldberg) [inline index.js:10]: I personally find it confusing to have `require`s nested within other statements. It'll also be harder to convert this to ESM if & when the time comes.

File: README.md [hunk 9]
```diff
@@ -1,83 +1,250 @@
...
+## Install
 
-**Help wanted:** We are looking for volunteers to maintain this project, if you are interested, feel free to contact me at [me@omnidan.net](mailto:me@omnidan.net)
+```sh
+npm install node-emoji
+```
 
-## Installation
-To install `node-emoji`, you need [node.js](http://nodejs.org/) and [npm](https://github.com/npm/npm#super-easy-install). :rocket:
+## Usage
```
Comment (JoshuaKGoldberg) [inline README.md:13]: Now that emojis are being sourced from other libraries, it'd probably be good to link to those libraries in the README.md.

File: package.json [hunk 10]
```diff
@@ -1,11 +1,23 @@
...
-    "type": "git",
-    "url": "https://github.com/omnidan/node-emoji.git"
+  "license": "MIT",
+  "main": "index.js",
+  "files": [
+    "index.js",
+    "index.d.ts"
+  ],
+  "engines": {
+    "node": ">=10"
```
Comment (JoshuaKGoldberg) [inline package.json:14]:
```suggestion
    "node": ">=14"
```

Comment (omnidan) [reply package.json:14]: any reason why we should support 14 and higher only?

Comment (JoshuaKGoldberg) [reply package.json:14]: Node 12 is now only in maintenance, and 14 is LTS. I suppose 12 would work too 🙂 but newer Node versions mean we can go with newer JS syntax.

File: .editorconfig (hunk 4 above)
Comment (JeDaYoshi) [reply .editorconfig:13]: I mean, I am personally up for it - so our editors know what way to follow.

File: package.json (hunk 10 above)
Comment (JeDaYoshi) [reply package.json:14]: I say better keep it on 10 unless it's really needed, or the new additions are just great enough to justify the compatibility breakage.

File: README.md [hunk 11]
```diff
@@ -1,83 +1,250 @@
...
+The input string to strip the emojis from.
+
+#### options
+
+Type: `object`
+
+##### removeSpaces
+
+Type: `boolean`\
+Default: `true`.
```
Comment (JoshuaKGoldberg) [inline README.md:172]: I've always found it confusing to have APIs that default a boolean parameter to `true` when not provided. Can we call this:

```suggestion
##### preserveSpaces

Type: `boolean`\
Default: `false`.
```

File: index.js [hunk 12]
```diff
@@ -1 +1,133 @@
...
+  assert.string(keyword)
+
+  keyword = normalizeName(keyword)
+
+  return emojiData
+    .filter(([name]) => name.includes(keyword))
+    .map(([name, emoji]) => ({ name, emoji }))
+}
+
+exports.find = emoji => {
```
Comment (JoshuaKGoldberg) [inline index.js:87]: Before there were `*ByCode` and `*ByName` equivalents for APIs such as `.find`. Was it intentional to remove them in v2? (I can't find discussion)

Comment (JoshuaKGoldberg) [reply index.js:87]: @omnidan thoughts?

Comment (omnidan) [reply index.js:87]: I would not know of a reason to remove them, maybe they were just forgotten about in the rewrite? /cc @Richienb

Comment (JoshuaKGoldberg) [reply index.js:87]: Added them back in #113.

**Pull request merged**